### DISCUSSION TITLE
How to find symmetry operations along a specific crystallographic direction in an FCC lattice?

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [spglib/spglib#503] [General]
Hi spglib developers,

I am currently working with a face-centered cubic (FCC) lattice and I need to identify the symmetry operations that preserve a specific crystallographic direction, specifically the [110] direction.

Here is the code I have been using:

```python
import spglib
import numpy as np

# Define the FCC lattice structure
a = 5.431
lattice = [[0.0, a/2, a/2],
           [a/2, 0.0, a/2],
           [a/2, a/2, 0.0]]
positions = [[0.0, 0.0, 0.0],  # Cu atom at origin
             [0.0, 0.5, 0.5],  # Cu atom at face center
             [0.5, 0.0, 0.5],  # Cu atom at face center
             [0.5, 0.5, 0.0]]  # Cu atom at face center
numbers = [1, 1, 1, 1]  # Use 1 to represent Cu atoms

cell = (lattice, positions, numbers)

# Get symmetry operations
dataset = spglib.get_symmetry_dataset(cell)
rotations = dataset['rotations']
translations = dataset['translations']

# Specify the crystallographic direction
direction = np.array([1, 1, 0])  # Along [110] direction

# Find symmetry operations that preserve the specified direction
symmetry_operations = []
for rotation, translation in zip(rotations, translations):
    transformed_direction = np.dot(rotation, direction)
    if np.allclose(transformed_direction, direction):
        symmetry_operations.append((rotation, translation))

# Output the results
print("Symmetry operations along the specified direction:")
for rotation, translation in symmetry_operations:
    print("Rotation:")
    print(rotation)
    print("Translation:")
    print(translation)
    print()
```

The result of the above code is as follows:

```
Symmetry operations along the specified direction:
Rotation:
[[1 0 0]
 [0 1 0]
 [0 0 1]]
Translation:
[0. 0. 0.]

Rotation:
[[0 1 0]
 [1 0 0]
 [0 0 1]]
Translation:
[0. 0. 0.]

Rotation:
[[1 0 0]
 [0 1 0]
 [0 0 1]]
Translation:
[0.5 0.  0.5]

Rotation:
[[0 1 0]
 [1 0 0]
 [0 0 1]]
Translation:
[0.5 0.  0.5]

Rotation:
[[1 0 0]
 [0 1 0]
 [0 0 1]]
Translation:
[0.5 0.5 0. ]

Rotation:
[[0 1 0]
 [1 0 0]
 [0 0 1]]
Translation:
[0.5 0.5 0. ]

Rotation:
[[1 0 0]
 [0 1 0]
 [0 0 1]]
Translation:
[0.  0.5 0.5]

Rotation:
[[0 1 0]
 [1 0 0]
 [0 0 1]]
Translation:
[0.  0.5 0.5]
```

However, I am not sure if this is the correct approach or if there are any specific functions in spglib that can directly provide the symmetry operations along a given direction.

Could you please provide guidance or suggestions on how to correctly identify these symmetry operations?

Thank you for your assistance!

Best regards,
Zhao

**Comment by atztogo** [comment]
Please confirm the below.

For FCC, you can choose either primitive cell or conventional unit cell. Your choice of `lattice` corresponds to the primitive cell. In this case, if you mean usual FCC (like Aluminum), there should be only one atom, i.e., I may write the script as follows,

```
a = 5.431
lattice = [[0.0, a/2, a/2],
           [a/2, 0.0, a/2],
           [a/2, a/2, 0.0]]
positions = [[0.0, 0.0, 0.0]]
numbers = [1]
```
What do you think?

**Reply by hongyi-zhao** [reply]
I agree with you, I give the results for both primitive cell and conventional cell cases:

1. For primitive cell: 
```python
import spglib
import numpy as np

# Define the FCC primitive cell lattice structure
a = 5.431
lattice = [[0.0, a/2, a/2],
           [a/2, 0.0, a/2],
           [a/2, a/2, 0.0]]
positions = [[0.0, 0.0, 0.0]]  # Cu atom at origin
numbers = [1]  # Use 1 to represent Cu atoms

cell = (lattice, positions, numbers)

# Get symmetry operations
dataset = spglib.get_symmetry_dataset(cell)
rotations = dataset['rotations']
translations = dataset['translations']

# Specify the crystallographic direction
direction = np.array([1, 1, 0])  # Along [110] direction

# Find symmetry operations that preserve the specified direction
symmetry_operations = []
for rotation, translation in zip(rotations, translations):
    transformed_direction = np.dot(rotation, direction)
    if np.allclose(transformed_direction, direction) or np.allclose(transformed_direction, -direction):
        symmetry_operations.append((rotation, translation))

# Output the results
print("Symmetry operations along the specified direction (Primitive Cell):")
for rotation, translation in symmetry_operations:
    print("Rotation:")
    print(rotation)
    print("Translation:")
    print(translation)
    print()
```

The result is as follows:

```
Symmetry operations along the specified direction (Primitive Cell):
Rotation:
[[1 0 0]
 [0 1 0]
 [0 0 1]]
Translation:
[0. 0. 0.]

Rotation:
[[-1  0  0]
 [ 0 -1  0]
 [ 0  0 -1]]
Translation:
[0. 0. 0.]

Rotation:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0 -1]]
Translation:
[0. 0. 0.]

Rotation:
[[0 1 0]
 [1 0 0]
 [0 0 1]]
Translation:
[0. 0. 0.]
```

2. For conventional cell:

```python
import spglib
import numpy as np

# Define the FCC conventional cell lattice structure
a = 5.431
lattice = [[a, 0.0, 0.0],
           [0.0, a, 0.0],
           [0.0, 0.0, a]]

# Define atomic positions for the conventional cell
positions = [[0.0, 0.0, 0.0],  # Corner
             [0.5, 0.5, 0.0],  # Face center
             [0.5, 0.0, 0.5],  # Face center
             [0.0, 0.5, 0.5]]  # Face center

numbers = [1, 1, 1, 1]  # Use 1 to represent Cu atoms

cell = (lattice, positions, numbers)

# Get symmetry operations
dataset = spglib.get_symmetry_dataset(cell)
rotations = dataset['rotations']
translations = dataset['translations']

# Specify the crystallographic direction
direction = np.array([1, 1, 0])  # Along [110] direction

# Find symmetry operations that preserve the specified direction
symmetry_operations = []
for rotation, translation in zip(rotations, translations):
    transformed_direction = np.dot(rotation, direction)
    if np.allclose(transformed_direction, direction) or np.allclose(transformed_direction, -direction):
        symmetry_operations.append((rotation, translation))

# Output the results
print("Symmetry operations along the specified direction (Conventional Cell):")
for rotation, translation in symmetry_operations:
    print("Rotation:")
    print(rotation)
    print("Translation:")
    print(translation)
    print()
```
The result is as follows:

```
Symmetry operations along the specified direction (Conventional Cell):
Rotation:
[[1 0 0]
 [0 1 0]
 [0 0 1]]
Translation:
[0. 0. 0.]

Rotation:
[[-1  0  0]
 [ 0 -1  0]
 [ 0  0 -1]]
Translation:
[0. 0. 0.]

Rotation:
[[-1  0  0]
 [ 0 -1  0]
 [ 0  0  1]]
Translation:
[0. 0. 0.]

Rotation:
[[ 1  0  0]
 [ 0  1  0]
 [ 0  0 -1]]
Translation:
[0. 0. 0.]

Rotation:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0 -1]]
Translation:
[0. 0. 0.]

Rotation:
[[0 1 0]
 [1 0 0]
 [0 0 1]]
Translation:
[0. 0. 0.]

Rotation:
[[ 0  1  0]
 [ 1  0  0]
 [ 0  0 -1]]
Translation:
[0. 0. 0.]

Rotation:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0  1]]
Translation:
[0. 0. 0.]

Rotation:
[[1 0 0]
 [0 1 0]
 [0 0 1]]
Translation:
[0.  0.5 0.5]

Rotation:
[[-1  0  0]
 [ 0 -1  0]
 [ 0  0 -1]]
Translation:
[0.  0.5 0.5]

Rotation:
[[-1  0  0]
 [ 0 -1  0]
 [ 0  0  1]]
Translation:
[0.  0.5 0.5]

Rotation:
[[ 1  0  0]
 [ 0  1  0]
 [ 0  0 -1]]
Translation:
[0.  0.5 0.5]

Rotation:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0 -1]]
Translation:
[0.  0.5 0.5]

Rotation:
[[0 1 0]
 [1 0 0]
 [0 0 1]]
Translation:
[0.  0.5 0.5]

Rotation:
[[ 0  1  0]
 [ 1  0  0]
 [ 0  0 -1]]
Translation:
[0.  0.5 0.5]

Rotation:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0  1]]
Translation:
[0.  0.5 0.5]

Rotation:
[[1 0 0]
 [0 1 0]
 [0 0 1]]
Translation:
[0.5 0.  0.5]

Rotation:
[[-1  0  0]
 [ 0 -1  0]
 [ 0  0 -1]]
Translation:
[0.5 0.  0.5]

Rotation:
[[-1  0  0]
 [ 0 -1  0]
 [ 0  0  1]]
Translation:
[0.5 0.  0.5]

Rotation:
[[ 1  0  0]
 [ 0  1  0]
 [ 0  0 -1]]
Translation:
[0.5 0.  0.5]

Rotation:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0 -1]]
Translation:
[0.5 0.  0.5]

Rotation:
[[0 1 0]
 [1 0 0]
 [0 0 1]]
Translation:
[0.5 0.  0.5]

Rotation:
[[ 0  1  0]
 [ 1  0  0]
 [ 0  0 -1]]
Translation:
[0.5 0.  0.5]

Rotation:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0  1]]
Translation:
[0.5 0.  0.5]

Rotation:
[[1 0 0]
 [0 1 0]
 [0 0 1]]
Translation:
[0.5 0.5 0. ]

Rotation:
[[-1  0  0]
 [ 0 -1  0]
 [ 0  0 -1]]
Translation:
[0.5 0.5 0. ]

Rotation:
[[-1  0  0]
 [ 0 -1  0]
 [ 0  0  1]]
Translation:
[0.5 0.5 0. ]

Rotation:
[[ 1  0  0]
 [ 0  1  0]
 [ 0  0 -1]]
Translation:
[0.5 0.5 0. ]

Rotation:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0 -1]]
Translation:
[0.5 0.5 0. ]

Rotation:
[[0 1 0]
 [1 0 0]
 [0 0 1]]
Translation:
[0.5 0.5 0. ]

Rotation:
[[ 0  1  0]
 [ 1  0  0]
 [ 0  0 -1]]
Translation:
[0.5 0.5 0. ]

Rotation:
[[ 0 -1  0]
 [-1  0  0]
 [ 0  0  1]]
Translation:
[0.5 0.5 0. ]
```

So, the above result is consistent with the one obtained [here](https://github.com/spglib/spglib/issues/502#issue-2393460216).

Now, I would like to know if the above results can be used to explain the following stress-strain phenomena as presented in the related literature [1-2]:

For an FCC crystal, it exhibits asymmetry when sheared along the positive and negative [11-2] directions, but it is symmetric when sheared along the positive and negative [110] directions. How can this be understood from the perspective of crystal symmetry and symmetry breaking after applying force?

[1] http://dx.doi.org/10.1016/j.actamat.2016.12.019
[1-s2.0-S1359645416309557-main.pdf](https://github.com/user-attachments/files/16115643/1-s2.0-S1359645416309557-main.pdf)
[2] https://www.science.org/doi/10.1126/science.adf1206
[science.adf1206.pdf](https://github.com/user-attachments/files/16115644/science.adf1206.pdf)

**Comment by atztogo** [comment]
I think your question is not about spglib, so I would not like to discuss it.

I want to avoid this kind of discussions here because it is inefficient to discuss before we are sure that we are standing on a similar ground. For example, it is unclear to me your definition about [110] or [11-2] directions, although I can guess. At least to discuss further, we have to have a common language, but which is not clear. I recommend you to read the standard text book, Physical Properties of Crystals - J. F. Nye, if you haven't read it.

**Comment by LecrisUT** [comment]
> However, I am not sure if this is the correct approach

Regarding the crystallographic aspect, I don't have the background or know the terminology differences. But to me the approach makes sense in a way, in that you take a vector normal to the plane you want and then check and make sure the rotation operators do not alter this vector. Whether the dot product does that and if it corresponds to the crystallographic notations I am not sure.

> if there are any specific functions in spglib that can directly provide the symmetry operations along a given direction.

There are none that I am aware of, but I don't see a reason why it shouldn't be on the high-level API levels. But the organization needs to be revised a bit first. See #489 for some ideas on that. Feel free to chime in or ping people you think would be interested in this, feedback early on such refactoring is very valuable. But if you want to discuss an interface for filtering symmetry operations like you have in this post, let's do it over here so that in that issue it is more towards the concept and interface.

**Reply by atztogo** [reply]
> But to me the approach makes sense in a way,

Even if it works, I could not judge if it is the desired result. I felt the question was not well defined, so the code can be fragile. Then, I thought the best solution is to recommend a book where fundamental knowledge is written very well. From my experience, these issues often come from the lack of basic knowledge, but not about how to use spglib.

**Reply by LecrisUT** [reply]
> Even if it works, I could not judge if it is the desired result. I felt the question was not well defined, so the code can be fragile. Then, I thought the best solution is to recommend a book where fundamental knowledge is written very well. From my experience, these issues often come from the lack of basic knowledge, but not about how to use spglib.

Sure, I don't judge one way or the other. Indeed having a more accurate description of the problem and its approach is preferred.

What I was considering instead was the suggestion of having ways to filter the symmetries. I can see that as a useful tool for the wider audience that do not have the full background of the crystallography. The details of the interface and implementation would need to be reviewed to avoid any pitfalls like you have mentioned with the directions of [110] or [11-2]. But I would encourage that if @hongyi-zhao can review the details and implementations, we should consider providing an interface for that (e.g. as a function `SpglibDataset.filter_symmetries()`)

**Reply by atztogo** [reply]
Thanks for your suggestion, @LecrisUT. 

The symmetry operations that spglib returns are a set of representations with respect to the basis vectors that the user chose. In this specific case, this point was unclear, either the user expects those of the primitive cell or conventional cell (or maybe something else), or if the user fully understands what he/she really wants.

I want to keep spglib as compact as possible. After obtaining the representations, other software package can easily implements such convenience features. So it is more reasonable to make other package for such convenient features. Providing uniform grid symmetry in reciprocal space is a good example.

**Reply by LecrisUT** [reply]
> I want to keep spglib as compact as possible. After obtaining the representations, other software package can easily implements such convenience features. So it is more reasonable to make other package for such convenient features. 

Fair concern. I agree with the sentiment to keep it as compact as possible. Things like constructing the reciprocal grid indeed feel out of scope for this project.

This filtering operation though feels like it can fit here. My reasoning is as follows:
- It seems like a basic operation that multiple projects would re-implement on their own
- For splitting such interface in a separate project, I don't know what project name or scope it would be in. Something like an `spglib-util` does not feel right to exist
- The operation and implementation looks very simple (once the details are resolved). This would be the main deciding factor for me for this proposal.
- The overall scope of `spglib` I would put it as:
  - Given a real space lattice and collection of atoms, what is the primitive cell
  - Provide basic crystallographic information and representations of the primitive cell
    Symmetry operations are among these. Filtering operations could go either way, similar to how the `SpglibDataset` provides a lot of baked-in representations.
    A keyword here would be `introspection` (interface).

> In this specific case, this point was unclear, either the user expects those of the primitive cell or conventional cell 

Good point. This feels like an issue to be addressed by OP when they implement the application.

**Reply by hongyi-zhao** [reply]
> But I would encourage that if @hongyi-zhao can review the details and implementations, we should consider providing an interface for that (e.g. as a function `SpglibDataset.filter_symmetries()`)

Frankly speaking, the attempt I'm making here is merely a potentially unreasonable speculation about the physical problem I'm trying to explain. In fact, I want to reasonably explain the following physical problem:

![image](https://github.com/spglib/spglib/assets/11155854/436360f9-9b80-405f-98b0-55decb48ab3b)
![image](https://github.com/spglib/spglib/assets/11155854/a438d5a9-d606-4bfc-9490-b54d5296d9e6)
![image](https://github.com/spglib/spglib/assets/11155854/5cb89e42-235d-4769-9bf2-c143c6cac41a)

I would like to understand why, under the application of shear stress in the positive and negative directions, the shear strain in figure A exhibits symmetrical changes, whereas in figure D it does not. I aim to use the symmetry of space groups to explain this issue, or at least provide some possible clues.

The figures above come from the literature mentioned before, aka, https://www.science.org/doi/10.1126/science.adf1206.

**Reply by LecrisUT** [reply]
> Frankly speaking, the attempt I'm making here is merely a potentially unreasonable speculation about the physical problem I'm trying to explain

That part we cannot help you here. Can try physics stackexchange or some other more general forums. Hopefully you can brainstorm something among your colleagues. Wish you the best of luck on your endeavors there.

What we can help you here is to improve the interface and documentation of this project. If you find some introspection interface to be essential we can discuss that in more detail.

**Reply by hongyi-zhao** [reply]
What do you mean by saying **introspection interface**?

**Reply by LecrisUT** [reply]
> What do you mean by saying **introspection interface**?

It's a general term for get, extract, select, filter, etc. For example you can introspect a python class to see what fields and methods it contains. In this context it's about introspecting `SpglibDataset` and what information is necessary to expose more readily.
